### PR TITLE
[Merged by Bors] - refactor: generalize `CharP+Prime` to `ExpChar` in `frobenius` 

### PR DESCRIPTION
**Pull request by alreadydone**
Consequently, the part about `frobenius` in Algebra/CharP/Basic is moved to CharP/ExpChar, and imports are adjusted as necessary.

+ Add instances from `CharP+Fact(Nat.Prime)` and `CharZero` to `ExpChar`, to allow lemmas generalized to ExpChar still apply to CharP.

+ Remove lemmas in Algebra/CharP/ExpChar from [#9799](https://github.com/leanprover-community/mathlib4/commit/1e74fcfff8d5ffe5a3a9881864cf10fa39f619e6) because they coincide with the generalized lemmas, and golf the other lemmas (in Algebra/CharP/Basic).

+ Define the RingHom `iterateFrobenius` and the semilinear map `LinearMap.(iterate)Frobenius` for an algebra. When the characteristic is zero (ExpChar is 1), these are all equal to the identity map (· ^ 1). Also define `iterateFrobeniusEquiv` for perfect rings.

+ Fix and/or generalize other files.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/CharP/ExpChar.lean [hunk 1]
```diff
@@ -244,46 +247,152 @@ theorem ExpChar.neg_one_pow_expChar_pow [Ring R] (q n : ℕ) [hR : ExpChar R q]
...
+  rfl
+#align frobenius_def frobenius_def
+
+theorem iterateFrobenius_def : iterateFrobenius R p n x = x ^ p ^ n := rfl
+
+theorem iterate_frobenius (n : ℕ) : (frobenius R p)^[n] x = x ^ p ^ n := by
+  induction n with
+  | zero => simp
+  | succ n n_ih =>
+      rw [Function.iterate_succ', pow_succ', pow_mul, Function.comp_apply, frobenius_def, n_ih]
```
Comment (acmepjz) [inline Mathlib/Algebra/CharP/ExpChar.lean:282]: Maybe you can add a result which states that `iterateFrobenius R p n` is equal to `(frobenius R p)^[n]` and this result is a corollary of it?

More generally, `iterateFrobenius R p (m * n)` is equal to `(iterateFrobenius R p m)^[n]`, similar to [Function.iterate_mul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Function/Iterate.html#Function.iterate_mul).

Also, `iterateFrobenius R p 1` is equal to `frobenius R p` and `iterateFrobenius R p 0` is equal to `RingHom.id`. Maybe these two should be simp lemmas?

Also, `iterateFrobenius R p (m + n)` is equal to `(iterateFrobenius R p m).comp (iterateFrobenius R p n)`.

---

Perhaps all of the above should be corollaries of corresponding results for `RingHom.iterate` (to be defined).

**Review by acmepjz — COMMENTED** [review]
Looks good to me.

File: Mathlib/Algebra/CharP/ExpChar.lean (hunk 1 above)
Comment (alreadydone) [reply Mathlib/Algebra/CharP/ExpChar.lean:282]: > Maybe you can add a result which states that `iterateFrobenius R p n` is equal to `(frobenius R p)^[n]` and this result is a corollary of it?

Good idea; that should be probably called `coe_iterateFrobenius`.

The other ideas are also good! Please feel free to add commits (and the same applies to #9311).

**Comment by acmepjz** [comment]
Maybe we should also replace the `CharP` in `Mathlib.FieldTheory.Perfect` by `ExpChar` as much as possible. Looks like that this will affect a lot of files, though.

**Comment by alreadydone** [comment]
You should not need to fix any file if you only replace CharP+Prime by ExpChar; I added an instance so that any declaration about ExpChar will continue to apply for CharP+Prime. If you replace CharP without an associated Nat.Prime assumption by ExpChar, then the declaration will be less general and may not apply in certain situations, and you'll need to keep the original lemma, but may still add an ExpChar version (for example `expand_contract'` in this PR).

Every CharP is associated with a Nat.Prime assumption in FieldTheory/Perfect though. However, there are declarations in the file that are tricky to generalize: for example `not_irreducible_expand` isn't true when p=1, and `PerfectRing (R × S) p` requires a `ExpChar (R × S) p` instance which in turn requires ` CharZero (R × S)` but both are missing. Moreover, even though a characteristic 0 field is a perfect field, I don't think people talk about characteristic 0 perfect rings (https://en.wikipedia.org/wiki/Perfect_ring is a different notion). So maybe there's not a good idea to generalize PerfectRing.

If you have specific lemmas that you want to generalize, you can try generalizing only those lemmas and their dependencies. This PR is probably long enough so you may consider doing it in another PR (maybe #9311).

**Comment by alreadydone** [comment]
In the end, I went ahead to generalize the whole file FieldTheory/Perfect, because it makes it smoother merging this into #9311.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**